src/command_modules/azure-cli-interactive/tests/test_query_injection.py

```diff
@@ -0,0 +1,136 @@
+        self.assertTrue(flag)
+        self.assertEqual('\n', self.stream.getvalue())
+
+    def test_just_query(self):
+        """ tests flushing just the query """
+        args = ['?x']
+        self.shell.last.result = {'x': 'result'}
+        flag = self.shell.handle_jmespath_query(args, False)
+        self.assertTrue(flag)
+        self.assertEqual('"result"\n', self.stream.getvalue())
+
+    def test_string_replacement(self):
+        """ tests that the query replaces the values in the command """
+        args = 'vm show -g ?group -n ?name'.split()
+        self.shell.last.result = {
+            'group': 'mygroup',
+            'name': 'myname'
+        }
+        flag = self.shell.handle_jmespath_query(args, False)
+        self.assertTrue(flag)
+        self.assertEqual(u"vm show -g mygroup -n myname\n", self.stream.getvalue())
+
+    def test_list_replacement(self):
+        """ tests that the query replaces the values in the command """
+        args = 'vm show -g ?[].group -n ?[].name'.split()
+        self.shell.last.result = [
+            {
+                'group': 'mygroup',
+                'name': 'myname'
+            },
+            {
+                'group': 'mygroup2',
+                'name': 'myname2'
+            },
+            {
+                'group': 'mygroup3',
+                'name': 'myname3'
+            }
+        ]
+        flag = self.shell.handle_jmespath_query(args, False)
+        self.assertTrue(flag)
+        results = self.stream.getvalue().strip().split('\n')
+        self.assertTrue(len(results) == 3)
+        self.assertEqual(results[0], 'vm show -g mygroup -n myname')
+        self.assertEqual(results[1], 'vm show -g mygroup2 -n myname2')
+        self.assertEqual(results[2], 'vm show -g mygroup3 -n myname3')
+
+    def test_quotes(self):
+        """ tests that it parses correctly with quotes in the command """
+        self.shell.last.result = {'x': 'result'}
+        # pylint: disable=protected-access
+        b_flag, c_flag, out, cmd = self.shell._special_cases(
+            "this is 'url?what' negative", "this is 'url?what' negative", False)
+        self.assertFalse(b_flag)
+        self.assertFalse(c_flag)
+        self.assertFalse(out)
+        self.assertEqual(cmd, "this is 'url?what' negative")
+
+    def test_errors(self):
+        """ tests invalid query """
+        args = 'vm show -g ?[].group -n ?[].name'.split()
+        self.shell.last.result = [
+            {
+                'group': 'mygroup',
+                'name': 'myname'
+            },
+            {
+                'group': 'mygroup3',
+            }
+        ]
+        flag = self.shell.handle_jmespath_query(args, False)
+        results = self.stream.getvalue().split('\n')
+        self.assertTrue(flag)
+        self.assertEqual(results[0], 'vm show -g mygroup -n myname')
+        self.assertEqual(
+            results[1], 'Invalid Input, the query gesture must have same number of items')
```

Comment:
Hmm... should this fail if this property (name, in the second example) doesn't exist, or should it just use None?  I'm not sure here, but it seems that this would be the equivalent of the service returning `'name': null` in the second instance and I'm not sure the error would make sense to the user.

Or, in this case, if you query and the thing isn't there, should it just skip that iteration? I'm not sure what the right answer should be here.